### PR TITLE
[JUJU-1368] Move more apis off the client facade

### PR DESCRIPTION
**Pull request by wallyworld**
Move more methods off the client facade.
The ModelUserInfo api is moved to the UserManager facade. This api needs the model uuid to gather users for that model. The UserManager facade is a controller facade so the model uuid needs to be passed in.

There were also methods to stream / download charms and blobs (in this case agent binaries). These methods are used by the migration worker, but they don't use a versioned api facade but rather perform an http GET against the underlying http connection. So we consolidate existing http code and move the generic blob streaming to `api/http` and move the charm download wrapper to the `api/charms` package.

Finally, the AgentVersion behaviour was wrong (because it reported the version of the running agent and not the model). It is replaced with a call to extract the model version from the model config. This is only used deploying local charms.

When moving the code, existing JujuConnSuite tests, where feasible, we rewritten using gomock.

## QA steps

bootstrap a controller, add a model, deploy some charms
juju users mymodel
juju users
check the above commands with an older controller

bootstrap another controller and migrate the model to it
after migration, add another unit to the model

### CODE REVIEW

File: api/client/charms/downloader.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/api/base"
+	"github.com/juju/juju/api/http"
+	"github.com/juju/juju/downloader"
+)
+
+// NewCharmDownloader returns a new charm downloader that wraps the
+// provided API caller.
+func NewCharmDownloader(apiCaller base.APICaller) *downloader.Downloader {
```
Comment (wallyworld) [inline api/client/charms/downloader.go:21]: Existing code, moved here, used by the unit agent when deploying a charm

File: api/client/client/client.go [hunk 2]
```diff
@@ -404,15 +346,6 @@ func (c *Client) httpPost(content io.ReadSeeker, endpoint, contentType string, r
 	return nil
 }
 
-// APIHostPorts returns a slice of network.MachineHostPort for each API server.
-func (c *Client) APIHostPorts() ([]network.MachineHostPorts, error) {
```
Comment (wallyworld) [inline api/client/client/client.go:408]: not used

File: api/http/http.go [hunk 3]
```diff
@@ -1,14 +1,71 @@
...
+
+// NewURIOpener returns a URI opener func for the api caller.
+func NewURIOpener(apiCaller base.APICaller) OpenURIFunc {
+	return func(uri string, query url.Values) (io.ReadCloser, error) {
+		return OpenURI(apiCaller, uri, query)
+	}
+}
+
+// OpenURI performs a GET on a Juju HTTP endpoint returning the specified blob.
+func OpenURI(apiCaller base.APICaller, uri string, query url.Values) (io.ReadCloser, error) {
```
Comment (tlm) [inline api/http/http.go:40]: I don't like this function signature.

This should take a context and a HTTPDoer not an apiCaller. It's a heavy object that makes testing the underlying premise hard and our reliance on gomock more.

File: api/http/http.go [hunk 4]
```diff
@@ -1,14 +1,71 @@
...
+	blob, err := openBlob(apiCaller.Context(), httpClient, uri, query)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return blob, nil
+}
+
+// openBlob streams the identified blob from the controller via the
+// provided HTTP client.
+func openBlob(ctx context.Context, httpClient HTTPDoer, endpoint string, args url.Values) (io.ReadCloser, error) {
```
Comment (tlm) [inline api/http/http.go:55]: This should take the url ready to go as a url.URL. This is a generic function that shouldn't be responsible for formatting the domain logic of something it doesn't understand.

File: api/http/http.go [hunk 5]
```diff
@@ -1,14 +1,71 @@
...
+	apiURL, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	apiURL.RawQuery = args.Encode()
+	req, err := http.NewRequest("GET", apiURL.String(), nil)
```
Comment (tlm) [inline api/http/http.go:61]: s/"GET"/http.MethodGet

File: api/http/http.go (hunk 3 above)
Comment (tlm) [inline api/http/http.go:40]: Should also take a well form URL see comments below.

File: api/http/http.go [hunk 6]
```diff
@@ -1,14 +1,71 @@
...
+}
+
+// OpenURI performs a GET on a Juju HTTP endpoint returning the specified blob.
+func OpenURI(apiCaller base.APICaller, uri string, query url.Values) (io.ReadCloser, error) {
+	// The returned httpClient sets the base url to /model/<uuid> if it can.
+	httpClient, err := apiCaller.HTTPClient()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	blob, err := openBlob(apiCaller.Context(), httpClient, uri, query)
```
Comment (tlm) [inline api/http/http.go:46]: nit, it's not a blob at this stage just reader to the blob. Is there a better name?

**Review by tlm — APPROVED** [review]

File: api/http/http.go (hunk 3 above)
Comment (wallyworld) [reply api/http/http.go:40]: done

File: api/http/http.go (hunk 5 above)
Comment (wallyworld) [reply api/http/http.go:61]: done

File: api/http/http.go (hunk 6 above)
Comment (wallyworld) [reply api/http/http.go:46]: done

File: api/client/charms/downloader.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+func NewCharmStreamer(apiCaller base.APICaller) CharmStreamerFunc {
+	return func(curl *charm.URL) (io.ReadCloser, error) {
+		uri, query := openCharmArgs(curl)
+		return http.OpenURI(apiCaller, uri, query)
+	}
+}
+
+// OpenCharm streams out the identified charm from the controller via
+// the API.
+func openCharm(apiCaller base.APICaller, curl *charm.URL) (io.ReadCloser, error) {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply api/client/charms/downloader.go:51]: done

File: api/client/charms/downloader.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+		},
+	}
+	return dlr
+}
+
+// CharmStreamerFunc returns a reader for the specified charm URL.
+type CharmStreamerFunc func(curl *charm.URL) (io.ReadCloser, error)
+
+// NewCharmStreamer returns a charm streamer func for the specified caller.
+func NewCharmStreamer(apiCaller base.APICaller) CharmStreamerFunc {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply api/client/charms/downloader.go:42]: This is called from places where we are passed an api caller and need to adapt it to become a charm opener. ie in the manifest startup. So it's appropriate to take an api caller and extract the context and underlying http client.

File: api/client/charms/downloader.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+// OpenCharm streams out the identified charm from the controller via
+// the API.
+func openCharm(apiCaller base.APICaller, curl *charm.URL) (io.ReadCloser, error) {
+	uri, query := openCharmArgs(curl)
+	return http.OpenURI(apiCaller, uri, query)
+}
+
+func openCharmArgs(curl *charm.URL) (string, url.Values) {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply api/client/charms/downloader.go:56]: The underlying juju infrastructure doesn't quite work that way.. See other comment.

File: api/http/http.go (hunk 3 above)
Comment (wallyworld) [reply api/http/http.go:40]: That's not what the juju infrastructure works expects in this case. What is passed in here is a relative path component which is added to the so called base URL which is derived from the api caller.

**Pull request merged**